crates/hir_ty/src/infer/expr.rs

```diff
@@ -430,11 +430,33 @@ impl<'a> InferenceContext<'a> {
                         name.as_tuple_index().and_then(|idx| substs.0.get(idx).cloned())
                     }
                     Ty::Adt(AdtId(hir_def::AdtId::StructId(s)), parameters) => {
-                        self.db.struct_data(s).variant_data.field(name).map(|local_id| {
-                            let field = FieldId { parent: s.into(), local_id };
-                            self.write_field_resolution(tgt_expr, field);
-                            self.db.field_types(s.into())[field.local_id].clone().subst(&parameters)
-                        })
+                        let def_db = self.db.upcast();
+                        let adt_module = hir_def::AdtId::from(s).module(def_db);
+                        let adt_resolver = adt_module.resolver(def_db);
+                        self.db.struct_data(s).variant_data.fields().iter().find_map(
+                            |(local_id, field_data)| {
+                                if &field_data.name != name {
+                                    return None;
+                                }
+                                let field = FieldId { parent: s.into(), local_id };
+                                self.write_field_resolution(tgt_expr, field);
+                                let field_vis =
+                                    field_data.visibility.resolve(def_db, &adt_resolver);
+                                if !self
+                                    .resolver
+                                    .module()
+                                    .map(|mod_id| field_vis.is_visible_from(def_db, mod_id))
+                                    .unwrap_or(false)
+                                {
+                                    return None;
+                                }
+                                Some(
+                                    self.db.field_types(s.into())[field.local_id]
+                                        .clone()
+                                        .subst(&parameters),
+                                )
+                            },
+                        )
```

Comment:
Is this the right approach?

Reply:
In principle yes, I think. This might be a bit nicer to write with some judicious use of `?` though :thinking:

Reply:
Will do.